### PR TITLE
application insights implementation

### PR DESCRIPTION
**Pull request by brettsam**
Tests are coming...

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs [hunk 1]
```diff
@@ -33,12 +35,14 @@ public class AdminController : ApiController
...
         private readonly TraceWriter _traceWriter;
+        private readonly ILogger _logger;
 
-        public AdminController(WebScriptHostManager scriptHostManager, WebHostSettings webHostSettings, TraceWriter traceWriter)
+        public AdminController(WebScriptHostManager scriptHostManager, WebHostSettings webHostSettings, TraceWriter traceWriter, ILoggerFactory loggerFactory)
         {
             _scriptHostManager = scriptHostManager;
             _webHostSettings = webHostSettings;
             _traceWriter = traceWriter.WithSource($"{ScriptConstants.TraceSourceHostAdmin}.Api");
+            _logger = loggerFactory?.CreateLogger("Host.Admin.Api");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: It looks like we're having the same duplicate logging pattern as in the other review. I thought you said you had a solution for that? It's a maintenance issue to have to remember to log double everywhere and ensure that the logs are the same

Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: Also, why are you using a string literal here instead of the constant used by the line above?

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs [hunk 2]
```diff
@@ -151,7 +156,18 @@ public static bool InStandbyMode
...
             {
-                await Instance.CallAsync(function.Name, arguments, cancellationToken);
+                // Add the request to the logging scope. This allows the App Insights logger to
+                // record details about the request.
+                ILoggerFactory loggerFactory = _config.HostConfig.GetService<ILoggerFactory>();
+                ILogger logger = loggerFactory.CreateLogger(LogCategories.Function);
+                using (logger.BeginScope(
+                    new Dictionary<string, object>()
+                    {
+                        ["MS_HttpRequest"] = request
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: nit: add this to our ScriptConstants

File: src/WebJobs.Script.WebHost/Controllers/AdminController.cs (hunk 1 above)
Comment (brettsam) [inline src/WebJobs.Script.WebHost/Controllers/AdminController.cs]: I do have a potential solution -- we've punted it until after this initial commit is in as it'll require a bunch of refactoring. Dispatching likely won't work very easily due to how ILoggers and TraceWriters are constructed (you have many ILoggers; few TraceWriters). But I'll experiment.

The string literal is temporary; I'll move it to a constant shortly. Looking for feedback on the right names, but that's not super-important for this initial commit as Categories, etc, are things that we can tweak during Preview. I can't use the existing ones because they don't follow the category 'namespacing' that I'm using for Loggers (i.e. everything from the host is 'Host.{category1}.{category2').

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 3]
```diff
@@ -266,7 +267,19 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
...
+                        // are correlated incorrectly. Here we'll overwrite that value with the correct value
+                        // immediately before logging.
+                        using (logger.BeginScope(
+                            new Dictionary<string, object>
+                            {
+                                ["MS_FunctionInvocationId"] = invocationContext.ExecutionContext.InvocationId
+                            }))
+                        {
+                            // TraceWriter already logs to ILogger
+                            traceWriter.Trace(evt);
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: I don't understand how the scope you're creating above affects this traceWriter log. This traceWriter isn't using app insights, right?

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 4]
```diff
@@ -266,7 +267,19 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
                     {
                         TraceLevel level = (TraceLevel)logData["lvl"];
                         var evt = new TraceEvent(level, message);
-                        traceWriter.Trace(evt);
+
+                        // Node captures the AsyncLocal value of the first invocation, which means that logs
+                        // are correlated incorrectly. Here we'll overwrite that value with the correct value
```
Comment (mathewc) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: You're saying we have a log correlation issue currently? Not sure I understand this.

File: src/WebJobs.Script/Diagnostics/FileLogger.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            IEnumerable<KeyValuePair<string, object>> properties = state as IEnumerable<KeyValuePair<string, object>>;
+            string formattedMessage = formatter?.Invoke(state, exception);
+
+            if (string.IsNullOrEmpty(formattedMessage) || properties == null || !IsEnabled(logLevel) || IsFromTraceWriter(properties))
+            {
+                return;
+            }
+
+            IDictionary<string, object> scopeProperties = DictionaryLoggerScope.GetMergedStateDictionary();
+            string functionName = scopeProperties["MS_FunctionName"].ToString();
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: looks like these constants are the same as you defined in core SDK scope keys. I think you should define all these in ScriptConstants.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs (hunk 4 above)
Comment (brettsam) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: This corrects the problem. Basically, node doesn't play well with AsyncLocal, so this corrects the scoping back to where it needs to be right before we write it out. @fabiocav and I discussed it a while back when I first ran into this.

File: src/WebJobs.Script/Diagnostics/FileLogger.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                functionName = functionName.Substring(functionsPrefix.Length);
+            }
+
+            TraceWriter traceWriter = _writerCache.GetOrAdd(functionName, (n) => CreateFileTraceWriter(n, _config));
+            TraceLevel traceLevel = GetTraceLevel(logLevel);
+            TraceEvent traceEvent = new TraceEvent(traceLevel, formattedMessage, _categoryName, exception);
+
+            // FileTraceWriter doesn't use these properties, but adding them for completeness
+
+            traceWriter.Trace(traceEvent);
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: Aren't we going to get duplicate logs? E.g. all over the place you had duplicate log statements one for our file traceWriter, and one for ILogger. The former will log to file already.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs (hunk 3 above)
Comment (brettsam) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: For logging within functions, we'll wire up an ILogger if one is registered: https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Bindings/TraceWriter/TraceWriterBinding.cs#L54-L59

This allows anyone currently using TraceWriter to automatically get App Insights logs if they enable them.

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 7]
```diff
@@ -46,6 +49,8 @@ internal BlobLeaseManager(ICloudBlob lockBlob, TimeSpan leaseTimeout, string hos
             _leaseRetryInterval = TimeSpan.FromSeconds(5);
 
             _timer = new Timer(ProcessLeaseTimerTick, null, TimeSpan.Zero, _leaseRetryInterval);
+
+            _logger = loggerFactory?.CreateLogger("Host.General");
```
Comment (mathewc) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: Please define constants like this in some central place so we can see them together and review for consistency

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 8]
```diff
@@ -226,7 +236,7 @@ internal static void AddFunctionError(Dictionary<string, Collection<string>> fun
             await CallAsync(methodInfo, arguments, cancellationToken);
         }
 
-        protected virtual void Initialize()
+        protected virtual void Initialize(ScriptSettingsManager settingsManager)
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: The instance already has a settings manager passed to it on construction, so we shouldn't need to modify this signature. Just use _settingsManager

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 9]
```diff
@@ -315,6 +325,13 @@ protected virtual void Initialize()
                     TraceWriter = new ConsoleTraceWriter(hostTraceLevel);
                 }
 
+                ConfigureLoggerFactory(settingsManager, metricsLogger);
+
+                // Use the startupLogger in this class as it is concerned with startup. The public Logger is used
+                // for all other logging after startup.
+                _startupLogger = ScriptConfig.HostConfig.LoggerFactory.CreateLogger(LogCategories.Startup);
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Some of your categories are defined in constants, but others are string literals like "Host.General"

Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Why do we have to switch on logger instance for different categories? Why isn't a category/scope for the logger used to do this instead?

File: src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        private const string Unknown = "[Unknown]";
+
+        public TelemetryClient Create(string instrumentationKey, SamplingPercentageEstimatorSettings samplingSettings)
+        {
+            ITelemetryClientFactory defaultFactory = new DefaultTelemetryClientFactory();
+            TelemetryClient client = defaultFactory.Create(instrumentationKey, samplingSettings);
+
+            string assemblyVersion = GetAssemblyFileVersion(typeof(ScriptHost).Assembly);
+            client.Context.GetInternalContext().SdkVersion = $"azurefuncs: {assemblyVersion}";
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs]: can this be "azurefunctions"?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 11]
```diff
@@ -1147,6 +1209,92 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
...
+
+                    if (aggregatorSection.TryGetValue("flushTimeout", out value))
+                    {
+                        scriptConfig.HostConfig.Aggregator.FlushTimeout = TimeSpan.Parse(value.ToString());
+                    }
+                }
+            }
+        }
+
+        internal static void ApplyApplicationInsightsConfig(JObject configJson, ScriptHostConfiguration scriptConfig)
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: You should have tests for this. For all our other config blocks, we have parsing unit tests (in ScriptHostTests I believe)

Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptHost.cs]: I've got tests for them -- I'll push them shortly.

File: src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs (hunk 10 above)
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs]: I'll double-check with App Insights. They recommended this (not sure if they have a limit on characters or anything).

File: src/WebJobs.Script/Diagnostics/FileLogger.cs [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Diagnostics
+{
+    /// <summary>
+    /// A wrapper class to allow ILoggers from functions to write to the existing FileTraceWriter. This allows
+    /// logs to show up in the log files and stream in the portal.
+    /// </summary>
+    internal class FileLogger : ILogger
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: We need tests for this new code, either unit or e2e somehow. Forgive me if I missed them, but I didn't see any tests for this stuff below.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 9 above)
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptHost.cs]: That's how ILogger works: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging#log-category

If we want categories to be meaningful, you need a new logger per category. The suggestion originally was to use one per-class (as the link above shows), but we decided that was even more confusing and harder to communicate to end-users when they look at their logs.

I'm open to suggestions -- we can tackle that in a subsequent checkin if we have a good solution. But this is how you define categories with ILogger.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 13]
```diff
@@ -1147,6 +1209,92 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
             {
                 scriptConfig.SwaggerEnabled = (bool)swaggerEnabled;
             }
+
+            ApplyLoggerConfig(config, scriptConfig);
+            ApplyApplicationInsightsConfig(config, scriptConfig);
+        }
+
+        internal static void ApplyLoggerConfig(JObject configJson, ScriptHostConfiguration scriptConfig)
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: You should have tests for this. For all our other config blocks, we have parsing unit tests (in ScriptHostTests I believe). E.g. ApplyConfiguration_Queues

File: src/WebJobs.Script/Diagnostics/FileLogger.cs (hunk 12 above)
Comment (brettsam) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: Well, my original PR said "tests coming soon" -- I forgot to put it into this PR. Yes, they're coming.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 13 above)
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptHost.cs]: They're coming!

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 14]
```diff
@@ -253,7 +254,7 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
             }
         }
 
-        private async Task<Dictionary<string, object>> CreateScriptExecutionContextAsync(object input, DataType dataType, TraceWriter traceWriter, FunctionInvocationContext invocationContext)
+        private async Task<Dictionary<string, object>> CreateScriptExecutionContextAsync(object input, DataType dataType, TraceWriter traceWriter, ILogger logger, FunctionInvocationContext invocationContext)
```
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: Why not just pull the ILogger from the FunctionInvocationContext, instead of adding a new parameter?

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs (hunk 3 above)
Comment (mamaso) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: Not a fan of this - tricky side effects & hard to follow.

I think I'd prefer if we didn't automatically wire up tracewriter in that way for functions - you'd need to add logger duplication but it would be clearer as to what's happening. Also, the AsyncLocal thing might not be an issue in that case? Not to clear on the issue there.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs (hunk 14 above)
Comment (brettsam) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: Hmm, good point. I guess the TraceWriter can't do that b/c it needs to be wrapped.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs (hunk 3 above)
Comment (brettsam) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: I mostly agree with you. We can revisit this -- at first I wanted to keep everything separate so it was clearer but I think both @paulbatum and @christopheranderson wanted users of TraceWriters to automatically get App Insights. And since this TraceWriter is coming from that same TraceWriterBinding, we'd need a way to "undo" that, which also gets messy.

I've got my fingers crossed that we can remove TraceWriter completely soon and then all of this confusion starts to go away -- but that's a separate discussion :-)

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 15]
```diff
@@ -243,12 +243,7 @@ protected virtual void Initialize(ScriptSettingsManager settingsManager)
             string debugSentinelFileName = Path.Combine(hostLogPath, ScriptConstants.DebugSentinelFileName);
             this.LastDebugNotify = File.GetLastWriteTime(debugSentinelFileName);
 
-            IMetricsLogger metricsLogger = ScriptConfig.HostConfig.GetService<IMetricsLogger>();
-            if (metricsLogger == null)
-            {
-                metricsLogger = new MetricsLogger();
-                ScriptConfig.HostConfig.AddService<IMetricsLogger>(metricsLogger);
-            }
+            IMetricsLogger metricsLogger = CreateMetricsLogger();
```
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptHost.cs]: FYI -- I chose to move this out to a separate method b/c the `Initialize` method had hit its cyclomatic complexity limit.

File: src/WebJobs.Script/Diagnostics/FileLogger.cs (hunk 5 above)
Comment (brettsam) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: Done.

File: src/WebJobs.Script.WebHost/WebScriptHostManager.cs (hunk 2 above)
Comment (brettsam) [inline src/WebJobs.Script.WebHost/WebScriptHostManager.cs]: Done.

File: src/WebJobs.Script/Diagnostics/FileLogger.cs (hunk 6 above)
Comment (brettsam) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: I filter above -- if this came from the user's TraceWriter, it doesn't make it through to here. If it's from the ILogger, it does. User logging is the one case where using the traceWriter will log to both ILogger and TraceWriter.

I mentioned it above -- I wanted to avoid this as well by have a clean separation between ILogger and TraceWriter. But from a user's perspective, now they will automatically get App Insights with their existing traces... which is nice, but leads to some complexity when doing things like this.

File: src/WebJobs.Script/Host/BlobLeaseManager.cs (hunk 7 above)
Comment (brettsam) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: done.

File: src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs [hunk 16]
```diff
@@ -266,7 +267,19 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
...
                         var evt = new TraceEvent(level, message);
-                        traceWriter.Trace(evt);
+
+                        // Node captures the AsyncLocal value of the first invocation, which means that logs
+                        // are correlated incorrectly. Here we'll overwrite that value with the correct value
+                        // immediately before logging.
+                        using (logger.BeginScope(
+                            new Dictionary<string, object>
+                            {
+                                ["MS_FunctionInvocationId"] = invocationContext.ExecutionContext.InvocationId
```
Comment (fabiocav) [inline src/WebJobs.Script/Description/Node/NodeFunctionInvoker.cs]: Costant?

File: src/WebJobs.Script/Diagnostics/DictionaryLoggerScope.cs [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            _state = state;
+            Parent = parent;
+        }
+
+        internal DictionaryLoggerScope Parent { get; private set; }
+
+        public static DictionaryLoggerScope Current
+        {
+            get
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/DictionaryLoggerScope.cs]: nit: Expression bodied getter and setter here

File: src/WebJobs.Script/Diagnostics/FileLogger.cs [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+                case LogLevel.Critical:
+                    return TraceLevel.Error;
+                case LogLevel.None:
+                    return TraceLevel.Off;
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private static bool IsFromTraceWriter(IEnumerable<KeyValuePair<string, object>> properties)
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/FileLogger.cs]: Should this method be renamed? The logic checks if this is a user log.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 8 above)
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: ⬆️ 👍

File: src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs (hunk 10 above)
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs]: As we discussed, let's set the application (or other appropriate property) to the host id so we can scope based on that.

File: src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+            return client;
+        }
+
+        internal static string GetAssemblyFileVersion(Assembly assembly)
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs]: Don't we already have logic that gets the appropriate runtime version somewhere (ends up in the Version property of the host)? Can we reuse that instead?

File: src/WebJobs.Script.WebHost/Security/SecretManager.cs [hunk 20]
```diff
@@ -173,7 +182,9 @@ protected virtual void Dispose(bool disposing)
 
             KeyOperationResult result = await AddOrUpdateSecretAsync(secretsType, keyScope, secretName, secret, secretsFactory);
 
-            _traceWriter.InfoFormat(Resources.TraceAddOrUpdateFunctionSecret, secretsType, secretName, keyScope ?? "host", result.Result);
+            string message = string.Format(Resources.TraceAddOrUpdateFunctionSecret, secretsType, secretName, keyScope ?? "host", result.Result);
+            _traceWriter.Info(message);
+            _logger?.LogDebug(message);
```
Comment (brettsam) [inline src/WebJobs.Script.WebHost/Security/SecretManager.cs]: It certainly seems like it :-)

File: src/WebJobs.Script/Description/DotNet/FunctionMetadataResolver.cs [hunk 21]
```diff
@@ -45,6 +47,7 @@ public sealed class FunctionMetadataResolver : MetadataReferenceResolver, IFunct
                 "System.Runtime",
                 "System.Threading.Tasks",
                 "Microsoft.CSharp",
+                typeof(ILoggerFactory).Assembly.Location,
```
Comment (brettsam) [inline src/WebJobs.Script/Description/DotNet/FunctionMetadataResolver.cs]: Done.

File: src/WebJobs.Script/Description/FunctionDescriptorProvider.cs [hunk 22]
```diff
@@ -106,6 +107,8 @@ public virtual bool TryCreate(FunctionMetadata functionMetadata, out FunctionDes
             // Add ExecutionContext to provide access to InvocationId, etc.
             parameters.Add(new ParameterDescriptor(ScriptConstants.SystemExecutionContextParameterName, typeof(ExecutionContext)));
 
+            parameters.Add(new ParameterDescriptor("_logger", typeof(ILogger)));
```
Comment (brettsam) [inline src/WebJobs.Script/Description/FunctionDescriptorProvider.cs]: Done.

File: src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs (hunk 19 above)
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs]: Yeah, good point. I copied this from SDK (which didn't have that) so it got duplicated. I'll switch it.

File: src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs (hunk 10 above)
Comment (brettsam) [inline src/WebJobs.Script/Host/ScriptTelemetryClientFactory.cs]: I agree, but I think that change can go into SDK (I'll have to try it out). I'll take a look at it in the next round of improvements.
